File: release/npm/lib/cache.js
"use strict";

const fs = require("node:fs");
const os = require("node:os");
const path = require("node:path");

/**
 * Resolves the root cache directory used to store downloaded scythe
 * binaries, honouring `SCYTHE_CACHE_DIR` and XDG conventions.
 *
 * @param {object} opts
 * @param {NodeJS.ProcessEnv} opts.env
 * @param {string} opts.platform - `process.platform` value.
 * @param {string} opts.homedir
 * @returns {string}
 */
function resolveCacheRoot({ env, platform, homedir }) {
  if (env.SCYTHE_CACHE_DIR) {
    return env.SCYTHE_CACHE_DIR;
  }
  if (platform === "win32") {
    const base = env.LOCALAPPDATA || path.join(homedir, "AppData", "Local");
    return path.join(base, "scythe", "cache");
  }
  if (env.XDG_CACHE_HOME) {
    return path.join(env.XDG_CACHE_HOME, "scythe");
  }
  return path.join(homedir, ".cache", "scythe");
}

/**
 * Resolves the cached binary path for a given version.
 *
 * @param {object} opts
 * @param {NodeJS.ProcessEnv} opts.env
 * @param {string} opts.platform
 * @param {string} opts.homedir
 * @param {string} opts.version
 * @param {string} opts.binaryName
 * @returns {string}
 */
function cachedBinaryPath({ env, platform, homedir, version, binaryName }) {
  return path.join(resolveCacheRoot({ env, platform, homedir: homedir ?? os.homedir() }), version, binaryName);
}

/**
 * Decides whether a cached path may be reused without re-downloading.
 *
 * Mere existence is not enough. Binaries published by this package are
 * renamed into place only after a full write and a checksum match, but a
 * cache directory can still hold a truncated leftover from an older version
 * of this installer that wrote straight to the final path. An empty file, a
 * non-file, or (on POSIX) a binary with no executable bit are all signs of
 * exactly that, and must trigger a fresh, verified download instead.
 *
 * @param {string} binaryPath
 * @param {string} platform - `process.platform` value.
 * @param {(path: string) => import("node:fs").Stats} [statSync]
 * @returns {boolean}
 */
function isUsableCachedBinary(binaryPath, platform, statSync = fs.statSync) {
  let stats;
  try {
    stats = statSync(binaryPath);
  } catch {
    return false;
  }
  if (!stats.isFile() || stats.size === 0) return false;
  if (platform === "win32") return true;
  return (stats.mode & 0o111) !== 0;
}

module.exports = { resolveCacheRoot, cachedBinaryPath, isUsableCachedBinary };

File: release/pypi/src/scythe_sql/cache.py
"""Cache directory resolution for the downloaded scythe binary."""

import os
from pathlib import Path


def resolve_cache_root(*, env: dict[str, str], sys_platform: str, home: Path) -> Path:
    """Resolves the root cache directory, honouring `SCYTHE_CACHE_DIR` and XDG conventions."""
    override = env.get("SCYTHE_CACHE_DIR")
    if override:
        return Path(override)

    if sys_platform.startswith("win"):
        base = env.get("LOCALAPPDATA")
        base_path = Path(base) if base else home / "AppData" / "Local"
        return base_path / "scythe" / "cache"

    xdg_cache_home = env.get("XDG_CACHE_HOME")
    if xdg_cache_home:
        return Path(xdg_cache_home) / "scythe"

    return home / ".cache" / "scythe"


def cached_binary_path(*, env: dict[str, str], sys_platform: str, home: Path, version: str, binary_name: str) -> Path:
    """Resolves the cached binary path for a given version."""
    return resolve_cache_root(env=env, sys_platform=sys_platform, home=home) / version / binary_name


def default_home() -> Path:
    """Returns the current user's home directory."""
    return Path(os.path.expanduser("~"))
